daemons/pacemakerd/pcmkd_subdaemons.c

```diff
@@ -127,44 +129,92 @@ pcmkd_cluster_connected(void)
-                    }
-                    pcmk_process_exit(&(pcmk_children[i]));
-                    continue;
-                default:
-                    crm_exit(CRM_EX_FATAL);
-                    break;  /* static analysis/noreturn */
+    static int next_child = 0;
+    int rc = child_liveness(&pcmk_children[next_child]);
+
+    /* xxx move to tracing before release */
+    crm_notice("%s[%lld] checked as %d",
+                           pcmk_children[next_child].name,
+                           (long long) PCMK__SPECIAL_PID_AS_0(
+                            pcmk_children[next_child].pid),
+                            rc);
+
+    switch (rc) {
+        case pcmk_rc_ok:
+            pcmk_children[next_child].check_count = 0;
+            next_child++;
+            subdaemon_check_progress = time(NULL);
+            break;
+        case pcmk_rc_ipc_pid_only: // This case: it was previously OK
+            pcmk_children[next_child].check_count++;
+            if (pcmk_children[next_child].check_count >= PCMK_PROCESS_CHECK_RETRIES) {
+                crm_err("%s[%lld] is unresponsive to ipc after %d tries but "
+                        "we found the pid so have it killed that we can restart",
+                        pcmk_children[next_child].name,
+                        (long long) PCMK__SPECIAL_PID_AS_0(
+                            pcmk_children[next_child].pid),
+                        pcmk_children[next_child].check_count);
+                stop_child(&pcmk_children[next_child], SIGKILL);
+                if (pcmk_children[next_child].respawn) {
+                    /* as long as the respawn-limit isn't reached
+                       give it another round of check retries
+                     */
+                    pcmk_children[next_child].check_count = 0;
+                }
+            } else {
+                if (pcmk_children[next_child].respawn) {
+                    /* as long as the respawn-limit isn't reached
+                       and we haven't run out of connect retries
+                       we account this as progress we are willing
+                       to tell to sbd
+                     */
+                    subdaemon_check_progress = time(NULL);
+                }
             }
-        }
-        /* at least one of the processes found at startup
-         * is still going, so keep this recurring timer around */
-        keep_tracking = TRUE;
+            /* go to the next child and see if
+               we can make progress there
+             */
+            next_child++;
+            break;
+        case pcmk_rc_ipc_unresponsive:
+            if (pcmk_children[next_child].respawn) {
+                crm_err("%s[%lld] terminated%s",
+                        pcmk_children[next_child].name,
+                        (long long) PCMK__SPECIAL_PID_AS_0(
+                            pcmk_children[next_child].pid),
+                        (rc == pcmk_rc_ipc_pid_only)? " as IPC server" : "");
```

Comment:
Now that we don't fall through we don't need this check or the similar one below

Reply:
oops